core/backend/src/CodeService.ts

```diff
@@ -0,0 +1,521 @@
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+
+import { CloudSqlite } from "@bentley/imodeljs-native";
+import { AccessToken, BentleyError, GuidString, IModelStatus, MarkRequired, Mutable } from "@itwin/core-bentley";
+import { CodeProps } from "@itwin/core-common";
+import { BriefcaseDb } from "./IModelDb";
+import { SettingObject } from "./workspace/Settings";
+
+/**
+ * A readonly index of all used and reserved Codes for an iTwin. The CodeIndex may be slightly out-of-date
+ * with the master copy in the cloud, but it should be periodically synchronized. Whenever codes are reserved/updated/deleted
+ * locally, this copy is always up-to-date as of those changes.
+ * @alpha
+ */
+export interface CodeIndex {
+  /**
+   * Find the next available value for the supplied `SequenceScope`.
+   * If the sequence is full (there are no available values), this will throw an exception with `errorId="SequenceFull"`
+   * @param from the sequence and scope to search
+   * @returns the next available CodeValue in the sequence.
+   */
+  findNextAvailable: (from: CodeService.SequenceScope) => CodeService.CodeValue;
+
+  /**
+   * Find the highest currently used value for the supplied `SequenceScope`
+   * @param from the sequence and scope to search
+   * @returns the highest used value, or undefined if no values have been used.
+   */
+  findHighestUsed: (from: CodeService.SequenceScope) => CodeService.CodeValue | undefined;
+
+  /** Determine whether a code is present in this CodeIndex by its Guid. */
+  isCodePresent: (guid: CodeService.CodeGuid) => boolean;
+
+  /** Get the data for a code in this CodeIndex by its Guid.
+   * @returns the data for the code or undefined if no code is present for the supplied Guid.
+   */
+  getCode: (guid: CodeService.CodeGuid) => CodeService.CodeEntry | undefined;
+
+  /** Look up a code by its Scope, Spec, and Value.
+   * @returns the Guid of the code, or undefined if not present.
+   */
+  findCode: (code: CodeService.ScopeSpecAndValue) => CodeService.CodeGuid | undefined;
+
+  /** Look up a code spec by its name
+   * @throws if the spec is not present.
+   */
+  getCodeSpec: (props: CodeService.CodeSpecName) => CodeService.NameAndJson;
+
+  /** Call a `CodeIterator` function for all codes in this index, optionally filtered by a `CodeFilter ` */
+  forAllCodes: (iter: CodeService.CodeIterator, filter?: CodeService.CodeFilter) => void;
+
+  /** Call an iterator function for all code specs in this index, optionally filtered by a `ValueFilter ` */
+  forAllCodeSpecs: (iter: CodeService.NameAndJsonIterator, filter?: CodeService.ValueFilter) => void;
+}
+
+/**
+ * The services for querying, reserving, updating, and deleting codes for a BriefcaseDb (available via `BriefcaseDb.codeService`) whenever it is opened for write access.
```

Comment:
I only see `IModelDb.codeService` and it's marked `@internal`. Do all IModelDb's have a CodeService, or is it specific to briefcases?

Reply:
If you are in an editing session with a CodeService implementation loaded (by setting `CodeService.createForBriefcase`) and you open a `BriefcaseDb` for write access, we call that function with the `BriefcaseDb` . If the iModel has a CodeService associated with it Codes are checked for new or updated elements.

Reply:
I'm not sure that answers my questions so I'll rephrase.
1. The code service is accessed via `IModelDb.codeService` yet this documentation refers to `BriefcaseDb.codeService`. Does it make sense for a non-briefcase to have a code service? If so the docs should refer to `IModelDb.codeService`, otherwise the property should move.
2. This documentation of an `alpha` interface refers to an `internal` property. Is `IModelDb.codeService` really supposed to be `internal`? If so, documentation shouldn't refer to it.

Reply:
1. The interface `onElementArg` in Element.ts is passed from native code to TS, and it has a `IModelDb` (because you can insert elements into standalone dbs.) So `codeService` must be a member of `IModelDb` even though it will only ever be defined on `BriefcaseDb`.
2. I changed the `codeservice` member to be @alpha this morning.

Reply:
```
  if (iModel.isBriefcaseDb())
    iModel.codeService?.verifyCode(arg.props);
```
?

Reply:
One option: move _codeService to BriefcaseDb and do
```
  if (iModel.isBriefcaseDb())
    iModel.codeService?.verifyCode(props);
```

Another option: keep IModelDb._codeService but only expose the getter on BriefcaseDb.